src/lib/urlHelper.js

```diff
@@ -12,6 +14,39 @@ export function getNewEntryUrl(collectionName, direct) {
   return getUrl(`/collections/${ collectionName }/entries/new`, direct);
 }
 
+// Unreserved chars from RFC3987.
+const uriChars = /[\w\-.~]/i;
+const ucsChars = /[\xA0-\u{D7FF}\u{F900}-\u{FDCF}\u{FDF0}-\u{FFEF}\u{10000}-\u{1FFFD}\u{20000}-\u{2FFFD}\u{30000}-\u{3FFFD}\u{40000}-\u{4FFFD}\u{50000}-\u{5FFFD}\u{60000}-\u{6FFFD}\u{70000}-\u{7FFFD}\u{80000}-\u{8FFFD}\u{90000}-\u{9FFFD}\u{A0000}-\u{AFFFD}\u{B0000}-\u{BFFFD}\u{C0000}-\u{CFFFD}\u{D0000}-\u{DFFFD}\u{E1000}-\u{EFFFD}]/u;
+export function sanitizeIRI(str, { replacement }) {
+  let result = "";
+  // We cannot use a `map` function here because `string.split()` splits things like emojis into surrogate pairs.
+  for (const char of str) {
+    if (uriChars.test(char) || ucsChars.test(char)) {
+      result += char;
+    } else {
+      result += replacement;
+    }
+  }
+  return result;
+}
+
+export function sanitizeSlug(str, { replacement = '-' }) {
+  if (!isString(str)) throw "`sanitizeSlug` only accepts strings as input.";
+  if (!isString(replacement)) throw "the `sanitizeSlug` replacement character must be a string.";
+  let slug = str;
```

Comment:
I'd still like to avoid mutations - we can use a couple of Lodash functions (`flow` and `partialRight`) to go immutable without the cruft:

```js
const sanitize = flow([
  partialRight(sanitizeIRI, { replacement }),
  partialRight(sanitizeFilename, { replacement }),
]);

const sanitizedSlug = sanitize(slug);

// After declaring replacement expressions:

const normalizedSlug = slug
  .replace(...)
  .replace(...);
```

Be sure to keep your source comments, those are helpful.